[Yamaha] Proper multi zone support and menu navigation

davidgraeff opened the pull request:
## New features
* Add menu navigation channels
* Add preset select channel
* Add play info/control channels
* Add system power channel
* Implement handleConfigurationUpdate to avoid the unnecessary dispose/initialize process.
* Add a workaround if the receiver returns HDMI$ instead of HDMI_$ as the input id to support more receiver models.
* Implement dynamic channel change for available inputs instead of listing all available inputs.
* Instead of the default preset channel names like Item_1..Item_40 the actual preset channel names are now obtained and the OH preset channel is dynamically updated

## Fixes
* Fixed several bugs related to zones
* Parsing exceptions of the xml parser are propagated through to the handler to allow a more helpful error message if parsing failed.

fixes #1701
fixes #1691

davidgraeff commented:
This pull request will incorporate changes from #1341

davidgraeff commented:
@kaikreuzer This can be merged if you have no further objections.

davidgraeff commented:
Ready to go test jar:

[org.openhab.binding.yamahareceiver_2.1.0.ref-rc1.jar.pdf](https://github.com/openhab/openhab2-addons/files/789530/org.openhab.binding.yamahareceiver_2.1.0.ref-rc1.jar.pdf)

davidgraeff commented:
Ready to test jar:

[org.openhab.binding.yamahareceiver_2.1.0.ref-rc2.jar.pdf](https://github.com/openhab/openhab2-addons/files/800915/org.openhab.binding.yamahareceiver_2.1.0.ref-rc2.jar.pdf)

Fixes bridge/zones marked as offline.

dfrommi commented:
Latest JAR introduced some new issues. 
- menu navigation is no longer working
- zone state update does no longer update `playback_channels#playback_station` and `navigation_channels#navigation_menu`

I think that's the culprit:
```
2017-02-25 19:29:32.864 [DEBUG] [eiver.handler.YamahaZoneThingHandler] - Input changed to NET RADIO
2017-02-25 19:29:32.869 [DEBUG] [eiver.handler.YamahaZoneThingHandler] - includeInputWithNavigationControl
2017-02-25 19:29:32.871 [DEBUG] [eiver.handler.YamahaZoneThingHandler] - !supportedInputs.contains(currentInputID): NET RADIO
2017-02-25 19:29:32.891 [DEBUG] [eiver.handler.YamahaZoneThingHandler] - includeInputWithPlaybackControl
2017-02-25 19:29:32.892 [DEBUG] [eiver.handler.YamahaZoneThingHandler] - !supportedInputs.contains(currentInputID): NET RADIO
```

The input is called `NET RADIO` on my AVR (space instead of underscore) and that's why menu navigation is disabled.

davidgraeff commented:
[org.openhab.binding.yamahareceiver_2.1.0.ref-rc3.jar.pdf](https://github.com/openhab/openhab2-addons/files/801585/org.openhab.binding.yamahareceiver_2.1.0.ref-rc3.jar.pdf)

I have changed the debug messages and replaced the underscore.

davidgraeff commented:
@dfrommi: Could you be a reviewer for the binding?

dfrommi commented:
Sure. I'll try to do it tomorrow.

dfrommi commented:
@davidgraeff Menu navigation is not working for me with your latest jar file:
```
2017-02-28 20:36:19.813 [ItemCommandEvent          ] - Item 'Yamaha_NetRadioMenu' received command Hinzugefügten Radiostationen/Capital Edinburgh
2017-02-28 20:36:19.893 [ItemStateChangedEvent     ] - Yamaha_NetRadioMenu changed from Hinzugefügten Radiostationen/Kiss FM to Hinzugefügten Radiostationen/Capital Edinburgh
2017-02-28 20:36:19.895 [hingStatusInfoChangedEvent] - 'yamahareceiver:zone:5f9ec1b3_ed59_1900_4530_00a0deb62041:Main_Zone' changed from ONLINE to OFFLINE (COMMUNICATION_ERROR): Server returned HTTP response code: 400 for URL: http://192.168.1.168:80/YamahaRemoteControl/ctrl
```

Let's try to resolve this one before I'll do the review.
Can you log the request when the response is an error?

davidgraeff commented:
To what channel is the item bound to? As there is no channel with the name netradiomenu anymore, it got replaced by a more general approach. I've never seen a 400 error that's really interesting. Yup the request should be logged when this error occurs.

dfrommi commented:
Ah sorry, that's the item configuration:
```
String Yamaha_NetRadioMenu "Net Radio Menu [%s]" {channel="yamahareceiver:zone:5f9ec1b3_ed59_1900_4530_00a0deb62041:Main_Zone:navigation_channels#navigation_menu"}
```

I can also debug it, if it can wait a few days.

dfrommi commented:
Did some debugging and got navigation working again. A few changes are required in `InputWithNavigationControl`.

First problem: Though the input is call `NET RADIO` when you select it as input, it is still `NET_RADIO` as an XML tag. A dirty way to fix it:
```java
    private String wrInput(String message) {
        return "<" + inputID.replaceAll(" ", "_") + ">" + message + "</" + inputID.replaceAll(" ", "_") + ">";
    }
```
But a proper distinction between input selection name and control ID would be better.

Next issue has already been fixed: the integer division problem of `pageCount` and `currentPage`. This is working for me:
```java
    private boolean selectItem(String name) throws IOException, ParserConfigurationException, SAXException {
        final int pageCount = (int) Math.ceil(((double) cache.Max_Line) / MAX_PER_PAGE);
        final int currentPage = (int) Math.floor(((double) (cache.Current_Line - 1)) / MAX_PER_PAGE);
...
```

davidgraeff commented:
Hm there is a distinction already: inputName and inputId. Id is what is supposed to be used for the XML tags.

davidgraeff commented:
There is apparently a third version that is used at some spots that does not use the underscore. I'll have a look later.

davidgraeff commented:
[org.openhab.binding.yamahareceiver_2.1.0.ref-rc4.jar.pdf](https://github.com/openhab/openhab2-addons/files/812889/org.openhab.binding.yamahareceiver_2.1.0.ref-rc4.jar.pdf)
Changelog:
* Convert spaces to underscores for input id
* Use math.floor for current page computation

Could you test this build?
Thanks, David

davidgraeff commented:
@kaikreuzer : Please pull this merge request. The zone support is broken in the current binding which confuses users. Additional new features like menu navigation have been tested by many users in the community forum and works for the majority. I'd like to fix any upcoming issues in subsequent, smaller commits, if there are any.

Thanks, David

kaikreuzer commented on addons/binding/org.openhab.binding.yamahareceiver/ESH-INF/thing/thing-types.xml:
```diff
@@ -4,21 +4,19 @@
         </channels>
         
+        <properties>
+        <property name="version">0.0</property>
```
indentation wrong.
The values do not seem to be relevant - you should rather remove those properties (you can still set them through code)

kaikreuzer commented on addons/binding/org.openhab.binding.yamahareceiver/ESH-INF/thing/thing-types.xml:
```diff
@@ -31,39 +29,122 @@
                 <default>60</default>
                 <advanced>true</advanced>
             </parameter>
+            <parameter name="TRACE_LOG" type="boolean" required="false">
```
No, please do not introduce such parameters.
openHAB has one logging framework and mechanisms to configure it - bindings should not start to come up with their own stuff.

kaikreuzer commented on addons/binding/org.openhab.binding.yamahareceiver/ESH-INF/thing/thing-types.xml:
```diff
@@ -31,39 +29,122 @@
+		<channel-groups>
+		  <channel-group typeId="zone_channels" id="zone_channels"/>
+		  <channel-group typeId="playback_channels" id="playback_channels"/>
+          <channel-group typeId="navigation_channels" id="navigation_channels"/>
```
check indentation

kaikreuzer commented on addons/binding/org.openhab.binding.yamahareceiver/ESH-INF/thing/thing-types.xml:
```diff
@@ -31,39 +29,122 @@
                 <advanced>true</advanced>
             </parameter>
+            <!-- 
+            <parameter name="NETRADIOMENU" type="text" required="false">
```
remove, if not required

kaikreuzer commented on addons/binding/org.openhab.binding.yamahareceiver/ESH-INF/thing/thing-types.xml:
```diff
@@ -115,6 +196,204 @@
         <item-type>Switch</item-type>
         <label>Mute</label>
         <description>Enable/Disable Mute on the AVR</description>
+        <category>SoundVolume_Mute</category>
```
I don't think this exists

kaikreuzer commented on addons/binding/org.openhab.binding.yamahareceiver/README.md:
```diff
@@ -11,6 +11,11 @@ by providing host and port.
+If your receiver is using menu-based net radio navigation, you can use this binding to
+select radio stations from a configured menu.
+
+TODO: New features like presets, menu navigation and play info/control have to be added to this readme!
```
so please do it

kaikreuzer commented on addons/binding/org.openhab.binding.yamahareceiver/implementation.md:
```diff
@@ -21,5 +21,151 @@ There is a second discovery service, the ZoneDiscoveryService which is created
 and used in createCommunicationObject() Main_Zone thing handler.
 For every other zone a discovery result will be generated and can be added by the user.
 
+## Protocol
```
add empty lines after every headline and before and after any code section, see https://www.eclipse.org/smarthome/documentation/development/bindings/docs.html

kaikreuzer commented on addons/binding/org.openhab.binding.yamahareceiver/src/main/java/org/openhab/binding/yamahareceiver/internal/YamahaReceiverHandlerFactory.java:
```diff
@@ -8,26 +8,34 @@
  * handlers.
  *
- * @author David Gräff - Initial contribution
+ * @author David Graeff <david.graeff@web.de>
```
no need for the e-mail address here - the original version was fine

davidgraeff replied:
What is the nowadays way of using the author annotations in the Oh2 addons codebase? Should I  just remove those entirely? Should I use the international version of my name (to allow people who want to contact me, to be able to type my name)? Git blame is superior in finding out what an author contributed, so should it be the name only without an email address? How does another developers contact previous authors to sort out code questions if there is now way of contacting him? (That is the reason for the author tag, isn't it?) Does the name should match the github name or the signoff name?

kaikreuzer commented on addons/binding/org.openhab.binding.yamahareceiver/src/main/java/org/openhab/binding/yamahareceiver/internal/protocol/InputWithNavigationControl.java:
```diff
@@ -0,0 +1,440 @@
+ * menu.goToPath(menuDir);
+ * menu.selectItem(stationName);
+ *
+ * @author Dennis Frommknecht
```
did you mention him in your sign-off?

kaikreuzer commented on addons/binding/org.openhab.binding.yamahareceiver/src/main/java/org/openhab/binding/yamahareceiver/internal/protocol/InputWithNavigationControl.java:
```diff
@@ -0,0 +1,440 @@
+ *
+ * @author Dennis Frommknecht
+ * @author David Graeff
+ * @since 2.0.0
```
remove

MHerbst commented on addons/binding/org.openhab.binding.yamahareceiver/src/main/java/org/openhab/binding/yamahareceiver/YamahaReceiverBindingConstants.java:
```diff
@@ -40,5 +83,13 @@
     public static final String CONFIG_HOST_NAME = "HOST";
     public static final String CONFIG_ZONE = "ZONE";
     public static final String CONFIG_RELVOLUMECHANGE = "RELVOLUMECHANGE";
+    public static final String CONFIG_TRACE_LOG = "TRACE_LOG";
```
Please remove own log handling (see Kai's comment above)

MHerbst commented on addons/binding/org.openhab.binding.yamahareceiver/src/main/java/org/openhab/binding/yamahareceiver/YamahaReceiverBindingConstants.java:
```diff
@@ -40,5 +83,13 @@
     public static final String CONFIG_ZONE = "ZONE";
     public static final String CONFIG_RELVOLUMECHANGE = "RELVOLUMECHANGE";
+    public static final String CONFIG_TRACE_LOG = "TRACE_LOG";
+    // public static final String CONFIG_NETRADIOMENU = "NETRADIOMENU";
```
Please remove

MHerbst commented on addons/binding/org.openhab.binding.yamahareceiver/src/main/java/org/openhab/binding/yamahareceiver/discovery/YamahaDiscoveryParticipant.java:
```diff
@@ -53,10 +53,15 @@ public DiscoveryResult createResult(RemoteDevice device) {
+        if (port == 8080) {
+            port = 80;
+        }
+
```
File contains some empty lines that are not necessary. Please check formatting here and also in other classes.

davidgraeff replied:
I generally agree, but empty lines to group code blocks is subject to the author. I checked this occurrence and I'm happy with that specific empty line to be honest.

MHerbst commented on addons/binding/org.openhab.binding.yamahareceiver/src/main/java/org/openhab/binding/yamahareceiver/handler/YamahaBridgeHandler.java:
```diff
@@ -0,0 +1,283 @@
+            return;
+        }
+
+        Object interval_config_o = thing.getConfiguration().get(YamahaReceiverBindingConstants.CONFIG_REFRESH);
```
To have a constant naming scheme please use camel casing instead of underlines. Please change the names of other variables, too.

MHerbst commented on addons/binding/org.openhab.binding.yamahareceiver/src/main/java/org/openhab/binding/yamahareceiver/handler/YamahaBridgeHandler.java:
```diff
@@ -0,0 +1,283 @@
+            state.invalidate();
+            return;
+        } catch (ParserConfigurationException e) {
+            e.printStackTrace();
```
Use logger instead of printing to stderr

davidgraeff replied:
It's unfortunately not that easy because some older firmware avrs send unexpected responses to some specific commands but all the rest of the functionality is still working. I don't want to put the device offline because of that (I actually implemented like this because a user had exactly this problem)

MHerbst commented on addons/binding/org.openhab.binding.yamahareceiver/src/main/java/org/openhab/binding/yamahareceiver/handler/YamahaBridgeHandler.java:
```diff
@@ -0,0 +1,283 @@
+        } catch (ParserConfigurationException e) {
+            e.printStackTrace();
+        } catch (SAXException e) {
+            e.printStackTrace();
```
Use logger instead of printing to stderr

MHerbst commented on addons/binding/org.openhab.binding.yamahareceiver/src/main/java/org/openhab/binding/yamahareceiver/handler/YamahaZoneThingHandler.java:
```diff
@@ -0,0 +1,606 @@
+                updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.CONFIGURATION_ERROR, e.getMessage());
+                return;
+            } catch (ParserConfigurationException e) {
+                e.printStackTrace();
```
Use logger instead of printing to stderr

MHerbst commented on addons/binding/org.openhab.binding.yamahareceiver/src/main/java/org/openhab/binding/yamahareceiver/handler/YamahaZoneThingHandler.java:
```diff
@@ -0,0 +1,606 @@
+            } catch (ParserConfigurationException e) {
+                e.printStackTrace();
+            } catch (SAXException e) {
+                e.printStackTrace();
```
Use logger instead of printing to stderr

MHerbst commented on addons/binding/org.openhab.binding.yamahareceiver/src/main/java/org/openhab/binding/yamahareceiver/handler/YamahaZoneThingHandler.java:
```diff
@@ -0,0 +1,606 @@
+        } catch (IOException e) {
+            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR, e.getMessage());
+        } catch (ParserConfigurationException e) {
+            e.printStackTrace();
```
Use logger instead of printing to stderr

MHerbst commented on addons/binding/org.openhab.binding.yamahareceiver/src/main/java/org/openhab/binding/yamahareceiver/handler/YamahaZoneThingHandler.java:
```diff
@@ -0,0 +1,606 @@
+        } catch (ParserConfigurationException e) {
+            e.printStackTrace();
+        } catch (SAXException e) {
+            e.printStackTrace();
```
Use logger instead of printing to stderr

MHerbst commented on addons/binding/org.openhab.binding.yamahareceiver/src/main/java/org/openhab/binding/yamahareceiver/handler/YamahaZoneThingHandler.java:
```diff
@@ -0,0 +1,606 @@
+            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR, e.getMessage());
+            return;
+        } catch (ParserConfigurationException e) {
+            e.printStackTrace();
```
Use logger instead of printing to stderr

MHerbst commented on addons/binding/org.openhab.binding.yamahareceiver/src/main/java/org/openhab/binding/yamahareceiver/handler/YamahaZoneThingHandler.java:
```diff
@@ -0,0 +1,606 @@
+     */
+    @Override
+    public void availableInputsChanged(AvailableInputState msg) {
+        // ChannelUID channelUID = new ChannelUID(thing.getUID(), YamahaReceiverBindingConstants.CHANNEL_GROUP_ZONE,
```
Remove unnecessary code

davidgraeff replied:
This is a todo and has to be done by another merge request. But it took a day to get this far, and I don't really want to remove this commented code right now, because it is partly doing what I wanted to archive.
Is there a developer documentation of how to dynamically change channel states that are statically defined in the xml?

MHerbst commented on addons/binding/org.openhab.binding.yamahareceiver/src/main/java/org/openhab/binding/yamahareceiver/handler/YamahaZoneThingHandler.java:
```diff
@@ -0,0 +1,606 @@
+    @Override
+    public void playControlUpdated(PlayControlState msg) {
+        playControlState = msg;
+        // TODO show preset name. The channel state options have to be changed for that
```
Please do it :-)

davidgraeff replied:
Is there a developer documentation of how to dynamically change channel states that are statically defined in the xml? As far as I have found out, it's not straight forward and other addons have plenty of boiler code to archive this. I'd like to avoid another hundreds of lines for this merge request to be honest.

MHerbst commented on addons/binding/org.openhab.binding.yamahareceiver/src/main/java/org/openhab/binding/yamahareceiver/internal/protocol/InputWithPlayControl.java:
```diff
@@ -0,0 +1,258 @@
+ * {@link InputWithPlayControl.PlayControlState} instead.
+ *
+ * @author David Graeff
+ * @since 2.0
```
remove

MHerbst commented on addons/binding/org.openhab.binding.yamahareceiver/src/main/java/org/openhab/binding/yamahareceiver/internal/protocol/SystemControl.java:
```diff
@@ -0,0 +1,142 @@
+ * No state will be saved in here, but in {@link SystemControl.State} instead.
+ *
+ * @author David Gräff <david.graeff@tu-dortmund.de>
+ * @since 2.1
```
remove

MHerbst commented on addons/binding/org.openhab.binding.yamahareceiver/src/main/java/org/openhab/binding/yamahareceiver/internal/protocol/ZoneControl.java:
```diff
@@ -0,0 +1,292 @@
+ * @author David Gräff <david.graeff@tu-dortmund.de>
+ * @author Eric Thill
+ * @author Ben Jones
+ * @since 2.0
```
remove

davidgraeff replied:
Is there a maven command to clear the author/version tags? In OH1 addons it was common to list authors and versions and I didn't find a documentation saying how annotations in the OH2 codebase should look like.

MHerbst commented on addons/binding/org.openhab.binding.yamahareceiver/src/main/java/org/openhab/binding/yamahareceiver/test/YamahaEmulatedHttpXMLServer.java:
```diff
@@ -0,0 +1,286 @@
+ * Emulates a Yamaha Receiver API endpoint (Http/XML on localhost:12121/YamahaRemoteControl/ctrl)
+ *
+ */
+public class YamahaEmulatedHttpXMLServer implements Runnable {
```
Wouldn't it be better to put it into a separate test project resp. bundle? For normal use this not needed.

davidgraeff replied:
As soon as I know to how archive that I'll do this. But it will take some time, so will not be realized for this merge request.

kaikreuzer replied:
@davidgraeff You can simply move all test classes into a separate source folder (src/test/java, which outputs to target/classes) within the same project. This will make sure that the packaged bundle won't include the test code (and I second @MHerbst that that code shouldn't be shipped with the binary).

davidgraeff replied:
I'm sorry I lost track. I moved the files, but at least eclipse does not recognise java packages in any other folder than src/main/java. Is this normal? @kaikreuzer

kaikreuzer replied:
Well, you have to declare the folder to be a source folder, see https://www.eclipse.org/forums/index.php?t=msg&th=164357&goto=520475&#msg_520475.

davidgraeff replied:
Thanks, found it

MHerbst commented on addons/binding/org.openhab.binding.yamahareceiver/src/main/java/org/openhab/binding/yamahareceiver/handler/YamahaBridgeHandler.java:
```diff
@@ -0,0 +1,283 @@
+            relativeVolumeChangeFactor = 0.5f;
+        }
+
+        Boolean configTraceLog = (Boolean) thing.getConfiguration()
```
Please remove own trace logic. You could enable the extended trace output depending on the log level.

davidgraeff replied:
Agreed. I found out about this feature of logging after I'd implemented it like it is.

davidgraeff replied:
But I've got a question. Is it allowed to write an own file with 'log' data based on the log level of logger. This is what I would do then.

MHerbst replied:
I haven't seen an explicit hint regarding additional log files. Especially for long trace output I would prefer to write an own log file instead of filling-up the standard log.

davidgraeff commented:
I will not have WiFi the next weeks so I'm not able to do the review process fixes. Can anybody else do this?

dfrommi commented on addons/binding/org.openhab.binding.yamahareceiver/src/main/java/org/openhab/binding/yamahareceiver/test/TestDiscovery.java:
```diff
@@ -0,0 +1,62 @@
+import org.openhab.binding.yamahareceiver.YamahaReceiverBindingConstants;
+import org.openhab.binding.yamahareceiver.discovery.YamahaDiscoveryParticipant;
+
+public class TestDiscovery extends AbstractDiscoveryService {
```
I'd prefer this to be disabled by default. It's quite confusing if you enable the binding and get an additional discovery result.
Alternatively you could get rid of the test server here and instead provide a wiremock configuration to mock behaviour. The receiver could be added manually using things file.
That should be easy enough for development without real receiver.

davidgraeff replied:
It is disabled by default, please check the OSGi XML service files where you can find enabled=false :)

piejanssens commented:
@davidgraeff Don't bail on the home stretch 😄  I think it's easiest and quickest if you did it yourself. I had a look at the comments and they all seem straightforward.

davidgraeff commented:
@piejanssens: I expect more to come :D but seriously I'm in new Zealand and in the process of changing over to the south island and cell reception is told to be aweful over there and data plans are expensive anyway here :)

piejanssens commented:
@davidgraeff Oh now I get it, enjoy your travel! 😉

davidgraeff commented on addons/binding/org.openhab.binding.yamahareceiver/README.md:
````diff
@@ -35,7 +40,6 @@ Items:
-     Number Yamaha_NetRadio  "Net Radio" <netRadio> 
 ```
 	 
 ### Manually linking
````
fixed

martinvw commented on addons/binding/org.openhab.binding.yamahareceiver/src/main/java/org/openhab/binding/yamahareceiver/internal/protocol/HttpXMLSendReceive.java:
```diff
@@ -0,0 +1,258 @@
+    Document xml(String message) throws IOException, ParserConfigurationException, SAXException {
+        String response = "<?xml version=\"1.0\" encoding=\"utf-8\"?>" + message;
+        DocumentBuilder db = dbf.newDocumentBuilder();
+        Document doc = db.parse(new InputSource(new StringReader(response)));
```
Unneeded local before return

davidgraeff replied:
done

martinvw commented on addons/binding/org.openhab.binding.yamahareceiver/src/main/java/org/openhab/binding/yamahareceiver/internal/protocol/HttpXMLSendReceive.java:
```diff
@@ -0,0 +1,258 @@
+    }
+
+    /**
+     * Post the given xml message and return the response as an xml document node.
```
To me this method does not look like it's posting anything it seems just to convert an xml string to a Document. If you add docblocks please make them useful and correct

davidgraeff replied:
Indeed. I'll check all comments for being up-to-date.

davidgraeff replied:
done

martinvw commented on addons/binding/org.openhab.binding.yamahareceiver/src/main/java/org/openhab/binding/yamahareceiver/internal/protocol/InputWithNavigationControl.java:
```diff
@@ -0,0 +1,439 @@
+            "Pandora", "SIRIUS", "Rhapsody", "iPod", "HD_RADIO");
+
+    public static final int MAX_PER_PAGE = 8;
+    private boolean use_alternative_back_to_home_cmd = false;
```
Naming conventions

davidgraeff replied:
done

martinvw commented on addons/binding/org.openhab.binding.yamahareceiver/src/main/java/org/openhab/binding/yamahareceiver/internal/protocol/InputWithNavigationControl.java:
```diff
@@ -0,0 +1,439 @@
+     * The current state of the navigation
+     */
+    public static class State {
+        public String Menu_Name = null;
```
Naming conventions, this ones start with Capitals

davidgraeff replied:
done

martinvw commented on addons/binding/org.openhab.binding.yamahareceiver/src/main/java/org/openhab/binding/yamahareceiver/internal/protocol/InputWithNavigationControl.java:
```diff
@@ -0,0 +1,439 @@
+            String s = "";
+            for (String item : items) {
+                if (item != null && item.length() > 0) {
+                    s += item + ",";
```
Please use a stringBuilder for this

davidgraeff replied:
done

martinvw commented on addons/binding/org.openhab.binding.yamahareceiver/src/main/java/org/openhab/binding/yamahareceiver/internal/protocol/InputWithNavigationControl.java:
```diff
@@ -0,0 +1,439 @@
+     * @param com The Yamaha communication object to send http requests.
+     */
+    public InputWithNavigationControl(String inputID, HttpXMLSendReceive com, Listener observer) {
+        this.com_ref = new WeakReference<HttpXMLSendReceive>(com);
```
Naming

davidgraeff replied:
done

martinvw commented on addons/binding/org.openhab.binding.yamahareceiver/src/main/java/org/openhab/binding/yamahareceiver/internal/protocol/InputWithNavigationControl.java:
```diff
@@ -0,0 +1,439 @@
+        if (pathArr.length < 2) {
+            if (!selectItem(pathArr[0])) {
+                observer.navigationError("Item '" + pathArr[0] + "' doesn't exist in menu " + getMenuName());
+                return;
```
Unneeded return also 251

davidgraeff replied:
done

martinvw commented on addons/binding/org.openhab.binding.yamahareceiver/src/main/java/org/openhab/binding/yamahareceiver/internal/protocol/InputWithNavigationControl.java:
```diff
@@ -0,0 +1,439 @@
+            try {
+                Thread.sleep(ZoneControl.MENU_RETRY_DELAY);
+            } catch (InterruptedException e) {
+                // Ignore and just retry immediately
```
If we receive this InterruptedException, the interrupted flag was cleared and we most likely need to reset the flag:

http://www.yegor256.com/2015/10/20/interrupted-exception.html
https://www.ibm.com/developerworks/library/j-jtp05236/

davidgraeff replied:
done

martinvw commented on addons/binding/org.openhab.binding.yamahareceiver/src/main/java/org/openhab/binding/yamahareceiver/internal/protocol/InputWithPlayControl.java:
```diff
@@ -0,0 +1,257 @@
+            "Napster", "Pandora", "SIRIUS", "Rhapsody", "Bluetooth", "iPod", "HD_RADIO");
+
+    public static class PlayInfoState {
+        public String Station; // NET_RADIO. Will also be used for TUNER where Radio_Text_A/B will be used instead.
```
Naming conventions

davidgraeff replied:
done

martinvw commented on addons/binding/org.openhab.binding.yamahareceiver/src/main/java/org/openhab/binding/yamahareceiver/internal/protocol/InputWithPlayControl.java:
```diff
@@ -0,0 +1,257 @@
+
+        if (inputID.equals("TUNER")) {
+            sub = HttpXMLSendReceive.getNode(metaInfoNode, "Radio_Text_A");
+            msg.Station = sub != null ? sub.getTextContent() : "";
```
Maybe you can group all sub == null together

davidgraeff replied:
Can you elaborate on that? Look at the line above those checks, sub is a different xml node each time. I didn't want to clutter the code with a new variable each time and I am reusing the 'sub' variable instead.

martinvw replied:
Okay that suggestion of me was incorrect :-S

You could extract a method:

```java
private String getNodeTextContent(Node metaInfoNode, String nodePath) {
    Node sub= HttpXMLSendReceive.getNode(metaInfoNode, "Radio_Text_A");
    return sub != null ? sub.getTextContent() : "";
}
```

davidgraeff replied:
should be all good like it is

martinvw commented on addons/binding/org.openhab.binding.yamahareceiver/src/main/java/org/openhab/binding/yamahareceiver/test/TestDiscovery.java:
```diff
@@ -0,0 +1,62 @@
+            DiscoveryResult discoveryResult = testClass.createResult(device);
+            thingDiscovered(discoveryResult);
+        } catch (Exception e) {
+            e.printStackTrace();
```
You should not use this, not even in this test-like class

davidgraeff replied:
done

martinvw commented on addons/binding/org.openhab.binding.yamahareceiver/src/main/java/org/openhab/binding/yamahareceiver/test/TestDiscovery.java:
```diff
@@ -0,0 +1,62 @@
+            RemoteDeviceIdentity identity = new RemoteDeviceIdentity(
+                    new UDN(UUID.fromString("e49825eb-8d1f-4e7a-8de4-c091850597f5")), 10, url, null,
+                    InetAddress.getByName(url.getHost()));
+            device = new RemoteDevice(identity, type, details, (RemoteService) null);
```
Should you group line 50 and 54?

davidgraeff replied:
done

martinvw commented on addons/binding/org.openhab.binding.yamahareceiver/src/main/java/org/openhab/binding/yamahareceiver/test/TestDiscovery.java:
```diff
@@ -0,0 +1,62 @@
+import org.jupnp.model.types.UDN;
+import org.openhab.binding.yamahareceiver.YamahaReceiverBindingConstants;
+import org.openhab.binding.yamahareceiver.discovery.YamahaDiscoveryParticipant;
+
```
Missing class header

davidgraeff replied:
done

martinvw commented on addons/binding/org.openhab.binding.yamahareceiver/src/main/java/org/openhab/binding/yamahareceiver/test/YamahaEmulatedHttpXMLServer.java:
```diff
@@ -0,0 +1,286 @@
+package org.openhab.binding.yamahareceiver.test;
+
```
It seems that this file has a missing  or invalid copyright header, can you fix all of them by running 

`mvn license:format` 

Please commit only the ones for your binding.

davidgraeff replied:
done

martinvw commented on addons/binding/org.openhab.binding.yamahareceiver/src/main/java/org/openhab/binding/yamahareceiver/test/YamahaEmulatedHttpXMLServer.java:
```diff
@@ -0,0 +1,286 @@
+    private final int port;
+
+    public static String parseBodyXML(InputStream inputStream, int len) throws IOException {
+        StringBuffer sb = new StringBuffer();
```
Please use the stringBuilder it's the modern alternative of this one

davidgraeff replied:
done

martinvw commented on addons/binding/org.openhab.binding.yamahareceiver/src/main/java/org/openhab/binding/yamahareceiver/test/YamahaEmulatedHttpXMLServer.java:
```diff
@@ -0,0 +1,286 @@
+            sb.append((char) charRead);
+        }
+
+        String data = sb.toString();
```
Unneeded local before return

davidgraeff replied:
done

martinvw commented on addons/binding/org.openhab.binding.yamahareceiver/src/main/java/org/openhab/binding/yamahareceiver/test/YamahaEmulatedHttpXMLServer.java:
```diff
@@ -0,0 +1,286 @@
+            }
+            server.close();
+        } catch (IOException e) {
+            e.printStackTrace();
```
PrintStackTrace

davidgraeff replied:
done

martinvw commented on addons/binding/org.openhab.binding.yamahareceiver/src/main/java/org/openhab/binding/yamahareceiver/test/YamahaEmulatedHttpXMLServer.java:
```diff
@@ -0,0 +1,286 @@
+            case "System": {
+                Node child = firstChild.getFirstChild();
+                if (!child.getTextContent().equals("GetParam")) {
+                    throw new Exception("Node for system/ only supports GetParam");
```
Do not throw exceptions of the class Exception use a specific applicable one.

davidgraeff replied:
done

martinvw commented on addons/binding/org.openhab.binding.yamahareceiver/src/main/java/org/openhab/binding/yamahareceiver/test/YamahaEmulatedHttpXMLServer.java:
```diff
@@ -0,0 +1,286 @@
+                        throw new Exception("Node for Input/Input_Sel_Item/ only supports GetParam");
+                    }
+                    return "<Input><Input_Sel_Item><Item1><Param>HDMI 1</Param><RW>RW</RW></Item1><Item2><Param>Net Radio</Param><RW>RW</RW></Item2></Input_Sel_Item></Input>";
+                }
```
Missing break, and you should not use braces in block of a switch

davidgraeff replied:
done

davidgraeff commented:
@martinvw Thank you for the review. I will incorporate the suggested changes and will have a another look at the coding guidelines as soon as I have some free time.

davidgraeff commented on addons/binding/org.openhab.binding.yamahareceiver/src/main/java/org/openhab/binding/yamahareceiver/handler/YamahaZoneThingHandler.java:
```diff
@@ -0,0 +1,606 @@
+                    logger.error(e.getMessage());
+                }
+            } else {
+                logger.debug("playback control  not supported by " + currentInputID);
```
double space

davidgraeff replied:
done

bartverstrynge commented:
Looking forward to test this!

piejanssens commented:
@davidgraeff Nearly ready?

davidgraeff commented:
@piejanssens You can use this binding already. I just don't have the abilities the next months to include the necessary changes to officially merge this.

bartverstrynge commented:
Would it be possible to add it to the store? I am eager to add this to my setup but I must be doing something wrong when trying to add it manually and this would allow us to install it consistently

piejanssens commented:
@davidgraeff Where can I find the latest test jar for this binding?

Weeves commented:
How does one get the rc jar?  The receiver zones are the next thing on my to-do list in my OH2 journey!

davidgraeff commented:
fixed all merge conflicts.

@Weeves : I do not provide a jar file for this anymore, there might be some outdated ones floating around in the forum.

bartverstrynge commented:
Also looking forward to use this!

On 8 Jun 2017 1:07 p.m., "David Gräff" <notifications@github.com> wrote:

> fixed all merge conflicts.
>
> @Weeves <https://github.com/weeves> : I do not provide a jar file for
> this anymore, there might be some outdated ones floating around in the
> forum.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/openhab/openhab2-addons/pull/1781#issuecomment-307072376>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ALjT3LZlR-G1ahMgrdidbeC78N02iMJmks5sB9YDgaJpZM4Lwbyd>
> .
>

davidgraeff commented on addons/binding/org.openhab.binding.yamahareceiver/src/main/java/org/openhab/binding/yamahareceiver/internal/protocol/HttpXMLSendReceive.java:
```diff
@@ -0,0 +1,258 @@
+    /**
+     * Set this flag to write out all send and received traffic
+     */
+    private boolean protocol_sniffer_enabled = false;
```
done

davidgraeff commented on addons/binding/org.openhab.binding.yamahareceiver/src/main/java/org/openhab/binding/yamahareceiver/internal/protocol/HttpXMLSendReceive.java:
```diff
@@ -0,0 +1,258 @@
+/**
+ * Copyright (c) 2014-2017 by the respective copyright holders.
```
done

davidgraeff commented on addons/binding/org.openhab.binding.yamahareceiver/src/main/java/org/openhab/binding/yamahareceiver/internal/YamahaReceiverHandlerFactory.java:
```diff
@@ -36,13 +44,21 @@ public boolean supportsThingType(ThingTypeUID thingTypeUID) {
     }
+
+    @Override
+    public Thing createThing(ThingTypeUID thingTypeUID, Configuration configuration, ThingUID thingUID,
```
done

davidgraeff commented on addons/binding/org.openhab.binding.yamahareceiver/src/main/java/org/openhab/binding/yamahareceiver/internal/YamahaReceiverHandlerFactory.java:
```diff
@@ -36,13 +44,21 @@ public boolean supportsThingType(ThingTypeUID thingTypeUID) {
+            return new YamahaZoneThingHandler(thing);
         }
 
+        logger.error("Yamaha thing factory  failed");
```
done

davidgraeff commented on addons/binding/org.openhab.binding.yamahareceiver/src/main/java/org/openhab/binding/yamahareceiver/handler/YamahaZoneThingHandler.java:
```diff
@@ -0,0 +1,606 @@
+            updateState(channelUID, new StringType(playInfoState.playbackMode));
+
+        } else if (id.equals(grpPlayback(YamahaReceiverBindingConstants.CHANNEL_PLAYBACK_STATION))) {
+            updateState(channelUID, new StringType(playInfoState.Station));
```
done

davidgraeff commented on addons/binding/org.openhab.binding.yamahareceiver/src/main/java/org/openhab/binding/yamahareceiver/handler/YamahaZoneThingHandler.java:
```diff
@@ -0,0 +1,606 @@
+
+                case YamahaReceiverBindingConstants.CHANNEL_NAVIGATION_MENU:
+                    if (inputWithNavigationControl == null) {
+                        logger.error("Channel " + id + " not working with this input!");
```
done

davidgraeff commented on addons/binding/org.openhab.binding.yamahareceiver/src/main/java/org/openhab/binding/yamahareceiver/handler/YamahaBridgeHandler.java:
```diff
@@ -0,0 +1,265 @@
+                updateState(channelUID, state.power ? OnOffType.ON : OnOffType.OFF);
+                break;
+            default:
+                logger.error("Channel refresh for " + channelUID.getId() + " not implemented!");
```
done

davidgraeff commented on addons/binding/org.openhab.binding.yamahareceiver/src/main/java/org/openhab/binding/yamahareceiver/handler/YamahaBridgeHandler.java:
```diff
@@ -0,0 +1,265 @@
+                }
+                    break;
+                default:
+                    logger.error("Channel " + id
```
done

davidgraeff commented on addons/binding/org.openhab.binding.yamahareceiver/src/main/java/org/openhab/binding/yamahareceiver/discovery/ZoneDiscoveryService.java:
```diff
@@ -21,17 +21,23 @@
-    public ZoneDiscoveryService() {
-        super(Collections.singleton(THING_TYPE_YAMAHAAV), 2, true);
+    public ZoneDiscoveryService(SystemControl.State state, ThingUID bridge_uid) {
+        super(YamahaReceiverBindingConstants.ZONE_THING_TYPES_UIDS, 2, true);
```
done

davidgraeff commented on addons/binding/org.openhab.binding.yamahareceiver/src/main/java/org/openhab/binding/yamahareceiver/handler/YamahaZoneThingHandler.java:
```diff
@@ -0,0 +1,606 @@
+            playInfoUpdated(playInfoState);
+            playInfoState.invalidate();
+        }
+
```
Method is split. But there is no good way of re-using code without introducing an interface for classes, who look similar but fulfill different purposes.

davidgraeff commented on addons/binding/org.openhab.binding.yamahareceiver/src/main/java/org/openhab/binding/yamahareceiver/handler/YamahaZoneThingHandler.java:
```diff
@@ -0,0 +1,606 @@
+                        brHandler.getCommunication(), this);
+                try {
+                    inputWithNavigationControl.updateNavigationState();
+                } catch (Exception e) {
```
done

davidgraeff commented on addons/binding/org.openhab.binding.yamahareceiver/src/main/java/org/openhab/binding/yamahareceiver/handler/YamahaZoneThingHandler.java:
```diff
@@ -0,0 +1,606 @@
+            zoneControl.updateState();
+        } catch (IOException e) {
+            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR, e.getMessage());
+            return;
```
it is necessary, because the thing will be put ONLINE otherwise at the end of the method. Only IO errors should cause the device to be OFFLINE.

martinvw replied:
Please check again there is no such statement in the end of this method :-)

davidgraeff commented on addons/binding/org.openhab.binding.yamahareceiver/src/main/java/org/openhab/binding/yamahareceiver/handler/YamahaZoneThingHandler.java:
```diff
@@ -0,0 +1,606 @@
+                    if (command instanceof PlayPauseType) {
+                        PlayPauseType t = ((PlayPauseType) command);
+                        switch (t) {
+                            case PAUSE:
```
Two nestings are fine with me in this case, and is somehow enforced by OH on me anyway. First I have to check for the channel ID and then for the command, which if it is an enum, is best processed by a switch. Personally I don't like to split the handleCommand() functionality.

davidgraeff commented on addons/binding/org.openhab.binding.yamahareceiver/src/main/java/org/openhab/binding/yamahareceiver/handler/YamahaZoneThingHandler.java:
```diff
@@ -0,0 +1,606 @@
+
+    private void initializeIfBridgeAvailable() {
+        // Do nothing if zone property is missing or we are already initialised
+        if (zone == null) {
```
done

davidgraeff commented on addons/binding/org.openhab.binding.yamahareceiver/src/main/java/org/openhab/binding/yamahareceiver/handler/YamahaZoneThingHandler.java:
```diff
@@ -0,0 +1,606 @@
+    public void handleConfigurationUpdate(Map<String, Object> configurationParameters) {
+        validateConfigurationParameters(configurationParameters);
+
+        // can be overridden by subclasses
```
done

davidgraeff commented on addons/binding/org.openhab.binding.yamahareceiver/src/main/java/org/openhab/binding/yamahareceiver/handler/YamahaZoneThingHandler.java:
```diff
@@ -0,0 +1,606 @@
+ * @author David Graeff <david.graeff@web.de>
+ */
+public class YamahaZoneThingHandler extends BaseThingHandler
+        implements Listener, org.openhab.binding.yamahareceiver.internal.protocol.InputWithNavigationControl.Listener,
```
done

davidgraeff commented on addons/binding/org.openhab.binding.yamahareceiver/src/main/java/org/openhab/binding/yamahareceiver/handler/YamahaZoneThingHandler.java:
```diff
@@ -0,0 +1,606 @@
+                updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.CONFIGURATION_ERROR, e.getMessage());
+                return;
+            } catch (ParserConfigurationException e) {
+                logger.error(e.getLocalizedMessage());
```
Yes, the device does not go offline if some responses cause parser errors. Some AVRs respond in a slightly different way, I want those errors to be "error" logged, because they will cause some functionality to not work.

davidgraeff commented on addons/binding/org.openhab.binding.yamahareceiver/src/main/java/org/openhab/binding/yamahareceiver/handler/YamahaBridgeHandler.java:
```diff
@@ -0,0 +1,265 @@
+            state.invalidate();
+            return;
+        } catch (ParserConfigurationException e) {
+            logger.error(e.getLocalizedMessage());
```
Yes

davidgraeff commented on addons/binding/org.openhab.binding.yamahareceiver/src/main/java/org/openhab/binding/yamahareceiver/handler/YamahaBridgeHandler.java:
```diff
@@ -0,0 +1,265 @@
+        // Read the configuration for the relative volume change factor.
+        BigDecimal relVolumeChange = (BigDecimal) thing.getConfiguration()
+                .get(YamahaReceiverBindingConstants.CONFIG_RELVOLUMECHANGE);
+        if (relVolumeChange != null) {
```
done

davidgraeff commented on addons/binding/org.openhab.binding.yamahareceiver/src/main/java/org/openhab/binding/yamahareceiver/handler/YamahaBridgeHandler.java:
```diff
@@ -0,0 +1,265 @@
+        if (refreshTimer != null) {
+            refreshTimer.cancel(false);
+        }
+        refreshTimer = scheduler.scheduleAtFixedRate(new Runnable() {
```
done

davidgraeff commented on addons/binding/org.openhab.binding.yamahareceiver/src/main/java/org/openhab/binding/yamahareceiver/handler/YamahaBridgeHandler.java:
```diff
@@ -0,0 +1,265 @@
+    }
+
+    private void refreshFromState(ChannelUID channelUID) {
+        switch (channelUID.getId()) {
```
added a comment for this design decision

davidgraeff commented on addons/binding/org.openhab.binding.yamahareceiver/src/main/java/org/openhab/binding/yamahareceiver/handler/YamahaBridgeHandler.java:
```diff
@@ -0,0 +1,265 @@
+                        // The user might have renamed some of the inputs etc.
+                        updateAllZoneInformation();
+                    }
+                }
```
done

davidgraeff commented on addons/binding/org.openhab.binding.yamahareceiver/src/main/java/org/openhab/binding/yamahareceiver/handler/YamahaBridgeHandler.java:
```diff
@@ -0,0 +1,265 @@
+                case YamahaReceiverBindingConstants.CHANNEL_POWER: {
+                    boolean oldState = state.power;
+                    SystemControl basicDeviceInformation = new SystemControl();
+                    basicDeviceInformation.setPower(xml, ((OnOffType) command) == OnOffType.ON, state);
```
I added documentation to basicDeviceInformation.setPower. That methods updates the state object if the command could successfully be transmitted to the device.

davidgraeff commented on addons/binding/org.openhab.binding.yamahareceiver/src/main/java/org/openhab/binding/yamahareceiver/handler/YamahaBridgeHandler.java:
```diff
@@ -0,0 +1,265 @@
+                return;
+            }
+
+            switch (id) {
```
Usually correct, but might be extended in the future. There is still unimplemented functionality. I added a comment.

davidgraeff commented:
@martinvw I think, I have incorporated all suggested changes. Could you have a look? I'll rebase/squash as soon as I get the OK for this merge.

martinvw commented on addons/binding/org.openhab.binding.yamahareceiver/src/main/java/org/openhab/binding/yamahareceiver/handler/YamahaZoneThingHandler.java:
```diff
@@ -0,0 +1,621 @@
+            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR, e.getMessage());
+            return;
+        } catch (ParserConfigurationException e) {
+            logger.error(e.getLocalizedMessage());
```
I might have missed something, [E.4 of the guidelines](http://docs.openhab.org/developers/development/guidelines.html#e-logging) describes log levels. Bindings should normally not log errors, please take a look at the page to check what would be proper here and adapt all `logger.error`'s

martinvw commented:
Thanks, I replied to one answer if you and added a remark about the logging levels. I'll also add this PR to @kaikreuzer 's queue.

davidgraeff commented:
I'm not logging to the error channel for response parser errors anymore and addressed the other issue with the unnecessary return.

martinvw commented:
OoO, travis has some issues with logging statements, sorry I did not notice that earlier:

```
ERROR] Failed to execute goal org.openhab.tools:static-code-analysis:0.0.6:report (default) on project org.openhab.binding.yamahareceiver: Code Analysis Tool has found 12 error(s)!
[ERROR] Please fix the errors and rerun the build.
[ERROR] Errors list:
[ERROR] ERROR found by findbugs: org.openhab.binding.yamahareceiver.handler.YamahaBridgeHandler.java:157 Format should be constant. Use placeholder to reduce the needless cost of parameter construction. see http://www.slf4j.org/faq.html#logging_performance
[ERROR] ERROR found by findbugs: org.openhab.binding.yamahareceiver.handler.YamahaZoneThingHandler.java:154 Format should be constant. Use placeholder to reduce the needless cost of parameter construction. see http://www.slf4j.org/faq.html#logging_performance
[ERROR] ERROR found by findbugs: org.openhab.binding.yamahareceiver.handler.YamahaZoneThingHandler.java:336 Format should be constant. Use placeholder to reduce the needless cost of parameter construction. see http://www.slf4j.org/faq.html#logging_performance
[ERROR] ERROR found by findbugs: org.openhab.binding.yamahareceiver.handler.YamahaZoneThingHandler.java:413 Format should be constant. Use placeholder to reduce the needless cost of parameter construction. see http://www.slf4j.org/faq.html#logging_performance
[ERROR] ERROR found by findbugs: org.openhab.binding.yamahareceiver.handler.YamahaZoneThingHandler.java:488 Format should be constant. Use placeholder to reduce the needless cost of parameter construction. see http://www.slf4j.org/faq.html#logging_performance
[ERROR] ERROR found by findbugs: org.openhab.binding.yamahareceiver.handler.YamahaZoneThingHandler.java:528 Format should be constant. Use placeholder to reduce the needless cost of parameter construction. see http://www.slf4j.org/faq.html#logging_performance
[ERROR] ERROR found by findbugs: org.openhab.binding.yamahareceiver.handler.YamahaZoneThingHandler.java:612 Format should be constant. Use placeholder to reduce the needless cost of parameter construction. see http://www.slf4j.org/faq.html#logging_performance
[ERROR] ERROR found by findbugs: org.openhab.binding.yamahareceiver.internal.protocol.HttpXMLSendReceive.java:89 Format should be constant. Use placeholder to reduce the needless cost of parameter construction. see http://www.slf4j.org/faq.html#logging_performance
[ERROR] ERROR found by findbugs: org.openhab.binding.yamahareceiver.internal.protocol.HttpXMLSendReceive.java:189 Format should be constant. Use placeholder to reduce the needless cost of parameter construction. see http://www.slf4j.org/faq.html#logging_performance
[ERROR] ERROR found by findbugs: org.openhab.binding.yamahareceiver.test.TestDiscovery.java:46 Format should be constant. Use placeholder to reduce the needless cost of parameter construction. see http://www.slf4j.org/faq.html#logging_performance
[ERROR] ERROR found by findbugs: org.openhab.binding.yamahareceiver.test.TestDiscovery.java:65 Format should be constant. Use placeholder to reduce the needless cost of parameter construction. see http://www.slf4j.org/faq.html#logging_performance
[ERROR] ERROR found by findbugs: org.openhab.binding.yamahareceiver.test.YamahaEmulatedHttpXMLServer.java:174 Format should be constant. Use placeholder to reduce the needless cost of parameter construction. see http://www.slf4j.org/faq.html#logging_performance
[ERROR] Detailed report can be found at: file:////home/travis/build/openhab/openhab2-addons/addons/binding/org.openhab.binding.yamahareceiver/target/code-analysis/report.html

```

davidgraeff commented:
Should be fixed now

martinvw commented:
Travis does not yet seem to be completely satisfied, can you please check?

martinvw commented:
There it is

[ERROR] ERROR found by findbugs: org.openhab.binding.yamahareceiver.handler.YamahaZoneThingHandler.java:612 Format should be constant. Use placeholder to reduce the needless cost of parameter construction. see http://www.slf4j.org/faq.html#logging_performance
[ERROR] Detailed report can be found at: file:////home/travis/build/openhab/openhab2-addons/addons/binding/org.openhab.binding.yamahareceiver/target/code-analysis/report.html
[ERROR] -> [Help 1]

davidgraeff commented:
findbugs is a little too picky in this last case, but i changed the code to make it happy

martinvw commented:
Thanks!

kaikreuzer commented:
@davidgraeff Could you check how to resolve the merge conflict?

@dfrommi As you seem to be actively using it - could you do some final test that all is working as expected?

davidgraeff commented:
Commits are squashed and merge conflict is fixed

davidgraeff commented:
Can we merge this soon please, I'm basically fixing merge conflicts all the time :D

bartverstrynge commented:
+1, I want my vacuum cleaner to start epic songs when it's finished

On 20 Jun 2017 9:54 a.m., "David Gräff" <notifications@github.com> wrote:

> Can we merge this soon please, I'm basically fixing merge conflicts all
> the time :D
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/openhab/openhab2-addons/pull/1781#issuecomment-309674152>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ALjT3KQZVB4Fgr-6n8jsaOtP4wYIwhaEks5sF3qxgaJpZM4Lwbyd>
> .
>

kaikreuzer commented:
@davidgraeff Did you see my https://github.com/openhab/openhab2-addons/pull/1781/files#r122004351? This does not seem to be addressed yet.

davidgraeff commented:
Tests packages are out of src/main/ now

kaikreuzer commented:
Another issue is that you still have `OSGI-INF/TestDiscovery.xml, which is packaged with the jar and which will cause exceptions at runtime as the classes cannot be found. Do you actually need this?
How are your tests executed at all? Sorry, but I do not quite yet have a clear view on this...

davidgraeff commented:
Test is the wrong wording. Those classes emulate a Yamaha receiver, based on some example responses derived from the Yamaha XML specification. It helped me to develop and test the binding and I intend to use this in my test binding for an osgi test. TestDiscovery.xml is disabled, will it still throw runtime errors?

kaikreuzer commented:
>  I intend to use this in my test binding for an osgi test

In this case you should definitely move all that stuff to a separate test fragment - we just have a new script that quickly helps you generating an according skeleton: 
https://github.com/openhab/openhab2-addons/blob/master/addons/binding/create_openhab_binding_test_skeleton.sh

davidgraeff commented:
Yup I know, I've send the PR. I need a way where I can turn this emulation on for the real binding, for development purposes. If it is in a test fragment, it won't be possible, is it?

kaikreuzer commented:
This is what test fragments are for: You can add them and have the code active or you can leave them away.

dfrommi commented:
I've done some testing based on  a JAR compiled from 21cb488. 

On/off and input switching is working, but Netradio navigation is not. I think it's again the space-vs-underscore problem (see my comment from Feb 25). Here the logs:
```
2017-06-24 13:57:31.708 [DEBUG] [eiver.handler.YamahaZoneThingHandler] - Input changed to NET_RADIO
2017-06-24 13:57:31.711 [DEBUG] [eiver.handler.YamahaZoneThingHandler] - navigation control not supported by NET_RADIO
2017-06-24 13:57:31.727 [DEBUG] [eiver.handler.YamahaZoneThingHandler] - playback control not supported by NET_RADIO
2017-06-24 13:57:37.161 [DEBUG] [eiver.handler.YamahaZoneThingHandler] - Zone state updated!
2017-06-24 13:57:37.164 [DEBUG] [eiver.handler.YamahaZoneThingHandler] - Input changed to HDMI_2
2017-06-24 13:57:37.250 [WARN ] [eiver.handler.YamahaZoneThingHandler] - post failed on: <?xml version="1.0" encoding="utf-8"?><YAMAHA_AV cmd="GET"><HDMI_2><List_Info>GetParam</List_Info></HDMI_2></YAMAHA_AV>
2017-06-24 13:57:37.294 [WARN ] [eiver.handler.YamahaZoneThingHandler] - post failed on: <?xml version="1.0" encoding="utf-8"?><YAMAHA_AV cmd="GET"><HDMI_2><Play_Info>GetParam</Play_Info></HDMI_2></YAMAHA_AV>
```
Netradio is activated and the default station loaded, but navigation is not working. I've then switched to HDMI2, which is also working, regardless of the 2 failed post commands.
It looks like you're trying to get list state for an HDMI input. At least on my receiver, there's no menu for HDMI input.

davidgraeff commented:
Looks like an inverted logic in the feature detection. The http posts should be issued for net radio and not for hdmi channels.   I'll have a look later.

davidgraeff commented:
@dfrommi The logic to determine if an input channel was capable of play control and navigation control was inverted. I've fixed this, could you have a look again?

dfrommi commented:
@davidgraeff I can now change netradio stations with `navigation_channels#navigation_menu` item, but there is a problem with `playback_channels#playback_station`. I see this in the logs:
```
2017-07-09 18:35:58.067 [ERROR] [eiver.handler.YamahaZoneThingHandler] - Channel playback_station not supported!
```
and the currently playing station is not updated.

davidgraeff commented:
@dfrommi This message indicates that you are trying to write/assign to the playback_station channel. It is read only and updated with all other playback channels as it should. There is a special case for TUNER to retrieve the current station though, maybe there need to be a special case for NET_RADIO as well. I don't have a Yamaha Receiver to test, so can somebody please test the **playback_station** channel for different inputs (DAB, TUNER, NET_RADIO)?

martinvw commented:
@kaikreuzer @davidgraeff is it clear to you two, who has to make the next step? I'm a little lost about the status :-)

zarusz commented:
In the meantime I started to test the Tuner DAB/FM support for this PR. My first impression is that there may be a small bug (working with @davidgraeff in priv mode). However, right now I am out of home for next 3 weeks and not able to test this through. That said, I could open a separate issue once the decision is to integrate this PR into master.

davidgraeff commented:
All raised issues are fixed, and if everything is reviewed, the PR is ready to be merged I guess

davidgraeff commented:
I really like to add support for https://community.openhab.org/t/yamaha-musiccast-binding/15060/30 (the new REST/JSON protocol of Yamaha, replacing the existing xml protocol). The architecture of the binding is already developed with this in mind, but if the merge takes longer (it is 6 month now since the initial PR), I'd like to at least know this. I will add the support on top of this PR then.

zarusz commented:
I managed to spend some time testing this new Yamaha binding version. Basically, changing presets does not work for dual band (FM/DAB+) yamaha receivers (RX-S601D in my case). Furthermore, there is still no way to change between bands FM or DAB+ when input is Tuner (this is a feature request).

Therefore, I created a separate issue #2524 to track the above mentioned and will contribute a new PR on top of this PR.

Please merge this PR when possible.

kaikreuzer commented:
Well, let's go for it then!

kaikreuzer commented on addons/binding/org.openhab.binding.yamahareceiver/ESH-INF/thing/thing-types.xml:
```diff
@@ -1,121 +1,384 @@
+		<description>Yamaha Receiver Zone</description>
+
+		<channel-groups>
+			<channel-group typeId="zone_channels" id="zone_channels" />
```
@davidgraeff I didn't notice it when I reviewed that PR, I only came across this right now: Why did you introduce channel-groups here? Those are meant for reusing the same set of channels multiple times, but you already chose to do this by Things. So I do not see any good reason to have channel groups here. Would it be possible to get rid off them?

davidgraeff replied:
The channel group usage is a leftover of a previous design, it works without them yes. A change touches some areas in the code and renders current configs invalid, I guess. I need to find out who is working on this add-on at the moment, there were some proposed changes in the community forum.
